### DISCUSSION TITLE
`uuid`s, what are they?

### DISCUSSION THREAD

**Discussion by lukors** [bevyengine/bevy#3995] [Q&A] [answered]
I'm looking at this example: https://github.com/bevyengine/bevy/blob/v0.6.1/examples/shader/shader_material.rs

At lines 47 through 51:

```rust
#[derive(Debug, Clone, TypeUuid)]
#[uuid = "4ee9c363-1124-4113-890e-199d81b00281"]
pub struct CustomMaterial {
    color: Color,
}
```

I  haven't been able to find any documentation about this for Bevy, so I'm wondering:

1. What's a `uuid`?
2. How do I know which `uuid` I should write?
3. Why is a `uuid` needed there in that example?

**Comment by bjorn3** [comment, marked as answer]
1. A uuid is a universally unique id. See https://en.wikipedia.org/wiki/Universally_unique_identifier for more info.
2. You should generate a random version 4 uuid. For example using https://www.uuidgenerator.net/ or by searching uuid on duckduckgo.
3. Materials are a kind of asset. Every asset type needs a unique uuid as asset handles consist of either a path (not applicable for runtime generated assets like this one) or the combination of a uuid corresponding to the asset type and a random 64bit integer. See https://github.com/bevyengine/bevy/blob/9a7852db0f22eb41f259a1afbb4926eb73863a10/crates/bevy_asset/src/handle.rs#L53-L55

**Reply by lukors** [reply]
Great explanation, thanks! :)

**Comment by Sorseg** [comment]
For anyone who, like me, still finds mentions of this in older documentation:
`TypeUuid` was [replaced](https://bevyengine.org/learn/migration-guides/0-12-to-0-13/#remove-typeuuid) with `TypePath` in bevy 0.13
